### ISSUE TITLE
Unclear manual for installation.

### ISSUE THREAD

**Issue by LWitter** [DeepLabCut/DeepLabCut#2183]
We were forced by our IT department to upgrade to win10 from win7. I dreaded this day because it meant reinstalling everything.
We did create yml files for our environments, but I also knew that getting CUDA, Cudnn, RTX drivers and Tensorflow to work together again could be an issue. (It is unfortunately)

So, I did a close read of the manual again, and noticed it could be clearer.
(this manual: https://github.com/DeepLabCut/DeepLabCut/blob/main/docs/installation.md)

Under "NVIDIA GPU? If you want to use your own GPU...." it is stated: " Note, DeepLabCut is up to date with the latest CUDA and tensorflow versions!"But further down, under "SECOND, install CUDA (versions up to CUDA11 are supported, together with TF2.5...." something else is implied, namely that CUDA and tensorflow are highly critically interdependent. Furthermore, tensorflow is up to version 2.11 when I'm writing this. No indication whether this would work.
To further complicate things it is stated that CudNN is provided with the environment from the github repository of DLC, but when looking at the Tensorflow site I see that also the version here is critically important.
Probably I am missing the details or nuance here, but as a relative outsider in using Python this information is highly confusing.

To make things worse, the order of things presented in the manual is not very intuitive (e.g. not an internal link to the GPU installation guide more down from "Step 1".) How to install CUDA for example is also not very clear. As a windows user it is simple to download from NVIDIA and install from the executable. But I probably need to install CUDA in anaconda, right? Maybe even in an environment? In that case, the whole order of things starts to break down. Again, highly confusing to me.

I would like to give very clear suggestions how to rewrite or how make more clear which steps are needed sequentially for non-knowledgeable users like me. When I have a better sense of how I finally got things to work (again) I will try to make a clear view of how (I think) I solved this.

**Comment by LWitter** [comment]
I got it to work, but it was quite haphazard I fear.

In the end these are the versions I installed (with CUDA being the latest, and no explicit install of Cudnn in Conda).
python		3.8.12
CUDA		10.1 <== pretty sure this is correct since there is only one folder with CUDA, but nvidia-smi says 12.1
GPU driver	latest: 531.41
pip		        23.0.1
Tensorflow	2.3.0
DLC		        2.2.0.3
ipython		7.30.1

Things clicked only after I removed all NVIDIA components and reinstalled CUDA 10.1. Then the driver failed and didn't install. Installing the latest finally solved my issues.

Again, I'm probably stupid in these things, but the above is how it seems to work now. I hope someone will find this small overview usefull. For completeness sake:
Windows 10
Xeon W2123
RTX2080Ti
(deeplabcut) C:\WINDOWS\system32>pip list
Package                       Version
----------------------------- -------------------
absl-py                       1.0.0
anipose                       1.0.1
aniposelib                    0.4.3
apptools                      5.1.0
argon2-cffi                   21.1.0
astunparse                    1.6.3
async-generator               1.10
attrs                         21.2.0
backcall                      0.2.0
backports.functools-lru-cache 1.6.4
bleach                        4.1.0
Brotli                        1.0.9
cachetools                    4.2.4
certifi                       2021.10.8
cffi                          1.15.0
charset-normalizer            2.0.9
click                         8.0.4
colorama                      0.4.4
configobj                     5.0.6
cycler                        0.11.0
debugpy                       1.5.1
decorator                     4.4.2
deeplabcut                    2.2.0.3
defusedxml                    0.7.1
entrypoints                   0.3
filterpy                      1.4.5
Flask                         2.0.3
Flask-Compress                1.11
flask-ipban                   1.1.5
flatbuffers                   2.0
fonttools                     4.28.3
gast                          0.3.3
google-auth                   2.3.3
google-auth-oauthlib          0.4.6
google-pasta                  0.2.0
grpcio                        1.42.0
h5py                          2.10.0
idna                          3.3
imageio                       2.13.3
imageio-ffmpeg                0.4.5
imgaug                        0.4.0
importlib-metadata            4.8.2
importlib-resources           5.4.0
ipykernel                     6.5.1
ipython                       7.30.1
ipython-genutils              0.2.0
ipywidgets                    7.6.5
itsdangerous                  2.1.1
jedi                          0.18.1
Jinja2                        3.0.3
joblib                        1.1.0
jsonschema                    4.2.1
jupyter                       1.0.0
jupyter-client                7.1.0
jupyter-console               6.4.0
jupyter-core                  4.9.1
jupyterlab-pygments           0.1.2
jupyterlab-widgets            1.0.2
keras                         2.8.0
Keras-Preprocessing           1.1.2
kiwisolver                    1.3.2
libclang                      12.0.0
llvmlite                      0.37.0
Markdown                      3.3.6
MarkupSafe                    2.0.1
matplotlib                    3.5.0
matplotlib-inline             0.1.3
mistune                       0.8.4
moviepy                       1.0.3
msgpack                       1.0.3
msgpack-numpy                 0.4.7.1
nb-conda                      2.2.1
nb-conda-kernels              2.3.1
nbclient                      0.5.9
nbconvert                     6.3.0
nbformat                      5.1.3
nest-asyncio                  1.5.4
networkx                      2.6.3
notebook                      6.4.6
numba                         0.54.1
numexpr                       2.8.0
numpy                         1.18.5
oauthlib                      3.1.1
opencv-contrib-python         4.5.5.64
opencv-python                 4.5.4.60
opencv-python-headless        4.5.4.60
opt-einsum                    3.3.0
packaging                     21.3
pandas                        1.3.4
pandocfilters                 1.5.0
parso                         0.8.3
patsy                         0.5.2
pickleshare                   0.7.5
Pillow                        8.4.0
pip                           23.0.1
proglog                       0.1.9
prometheus-client             0.12.0
prompt-toolkit                3.0.24
protobuf                      3.19.1
psutil                        5.8.0
pyasn1                        0.4.8
pyasn1-modules                0.2.8
pycparser                     2.21
pyface                        7.4.1
Pygments                      2.10.0
pyparsing                     3.0.6
PyQt5                         5.12.3
PyQt5_sip                     4.19.18
PyQtChart                     5.12
PyQtWebEngine                 5.12.1
pyrsistent                    0.18.0
python-dateutil               2.8.2
pytz                          2021.3
PyWavelets                    1.2.0
pywin32                       302
pywinpty                      1.1.6
PyYAML                        6.0
pyzmq                         22.3.0
qtconsole                     5.2.1
QtPy                          1.11.3
requests                      2.26.0
requests-oauthlib             1.3.0
rsa                           4.8
ruamel.yaml                   0.17.17
ruamel.yaml.clib              0.2.6
scikit-image                  0.18.1
scikit-learn                  1.0.1
scikit-video                  1.1.11
scipy                         1.4.1
Send2Trash                    1.8.0
setuptools                    59.4.0
setuptools-scm                6.3.2
Shapely                       1.8.0
six                           1.16.0
statsmodels                   0.13.1
tables                        3.6.1
tabulate                      0.8.9
tensorboard                   2.8.0
tensorboard-data-server       0.6.1
tensorboard-plugin-wit        1.8.0
tensorflow                    2.3.0
tensorflow-estimator          2.3.0
tensorflow-io-gcs-filesystem  0.24.0
tensorpack                    0.11
termcolor                     1.1.0
terminado                     0.12.1
testpath                      0.5.0
tf-estimator-nightly          2.8.0.dev2021122109
tf-slim                       1.1.0
threadpoolctl                 3.0.0
tifffile                      2021.11.2
toml                          0.10.2
tomli                         1.2.2
tornado                       6.1
tqdm                          4.62.3
traitlets                     5.1.1
traits                        6.3.2
traitsui                      7.3.1
typing_extensions             4.0.1
urllib3                       1.26.7
wcwidth                       0.2.5
webencodings                  0.5.1
Werkzeug                      2.0.2
wheel                         0.37.0
widgetsnbextension            3.5.2
wrapt                         1.13.3
wxPython                      4.1.1
zipp                          3.6.0

(deeplabcut) C:\WINDOWS\system32>nvidia-smi
Mon Mar 27 17:10:43 2023
+---------------------------------------------------------------------------------------+
| NVIDIA-SMI 531.41                 Driver Version: 531.41       CUDA Version: 12.1     |
|-----------------------------------------+----------------------+----------------------+
| GPU  Name                      TCC/WDDM | Bus-Id        Disp.A | Volatile Uncorr. ECC |
| Fan  Temp  Perf            Pwr:Usage/Cap|         Memory-Usage | GPU-Util  Compute M. |
|                                         |                      |               MIG M. |
|=========================================+======================+======================|
|   0  NVIDIA GeForce RTX 2080 Ti    WDDM | 00000000:65:00.0  On |                  N/A |
| 30%   31C    P8               24W / 260W|    218MiB / 11264MiB |      1%      Default |
|                                         |                      |                  N/A |
+-----------------------------------------+----------------------+----------------------+

+---------------------------------------------------------------------------------------+
| Processes:                                                                            |
|  GPU   GI   CI        PID   Type   Process name                            GPU Memory |
|        ID   ID                                                             Usage      |
|=======================================================================================|
|    0   N/A  N/A      4520    C+G   ....Search_cw5n1h2txyewy\SearchApp.exe    N/A      |
|    0   N/A  N/A      6048    C+G   ...CBS_cw5n1h2txyewy\TextInputHost.exe    N/A      |
|    0   N/A  N/A      8464    C+G   C:\Windows\explorer.exe                   N/A      |
|    0   N/A  N/A      9344    C+G   ...5n1h2txyewy\ShellExperienceHost.exe    N/A      |
+---------------------------------------------------------------------------------------+

(deeplabcut) C:\WINDOWS\system32>

**Comment by MMathisLab** [comment]
Hi @LWitter did you check out the nature protocols guide as a first pass? And no CUDA, driver and GPU are not installed inside anaconda. This is either for Docker or for your system. You can also watch the youtube videos for additional guidance. 

Also, thanks for the feedback, we will take it into account when we update docs, but please do note this is a free code project, without support or guarantees ;).

**Comment by LWitter** [comment]
I much appreciate the program and the support. Just wanted to make sure to communicate where I found the manual confusing.
I didn't know about the YouTube videos yet, will definitely check those out as well!

**Comment by jonahpearl** [comment]
Just echoing the request for an updated set of GPU related docs -- I struggled for a while to get things going on a Windows 10 machine for similar reasons. Even though I had the correct versions of CUDA and cudnn installed on my machine (nvidia-smi worked, nvcc -V was good) and *then* installed the DLC conda env (the envs was created fine and packages deeplabcut/torch/tensorflow imported with no warnings in a python session), when I ran the testscript, it was clear that tensorflow couldn't find the GPU. I double-checked sys paths, etc. Eventually I had to use conda to install them as described [here](https://github.com/DeepLabCut/DeepLabCut/issues/2115#issuecomment-1408298351), then tensorflow could see the GPU. So there's something weird about how conda envs are interacting with system-level installs of CUDA and cudnn. Maybe this is a known issue with conda, but I think the DLC docs should mention that if you're having trouble with system level installs, to try using conda to install them (rather than just pointing to the NVIDIA docs, which only describe system level installs). Thanks!

**Comment by linussun** [comment]
This post is for new users trying to get their NVIDIA GPUs to work with DLC on Windows 11. 

Also echoing  @LWitter @jonahpearl comments. Appreciate this resource, and will leave a comment here where the documentation didn't quite pinpoint my GPU installation problem in Windows 11 which was that cuDNN was **NOT** installed properly like the documentation and comments in Github state. Installing cuDNN to my DLC environment solved the problem of "python testscript.py" (another thing, for some reason pythonw doesn't work, but python does) failing due to tensorflow not being able to find cuDNN v11 libraries. Here was my step by step installation that resolved my setup's problem:

Hardware: Alienware M15 Laptop with an Nvidia 4080 Ti GPU running Windows 11

1. Nvidia website install GPU Drivers 531.41
2. Nvidia CUDA Install (version 12.1) - confirm with terminal command line commands "nvidia-smi" and "nvcc --version"
3.  Install Anaconda Navigator (2.4.0 for my setup; using python 3.8 in the base installation; always run in administrator mode).
4. Make sure git is installed in the base installation (can use Anaconda Navigator to do this) 
a. In an Anaconda3 terminal (run as admin) change the current working directory to your directory of choice and call to download the repository: "git clone https://github.com/DeepLabCut/DeepLabCut.git”  
b. "conda env create -f DEEPLABCUT.yaml"
5. I used Anaconda Navigator to install and verify the following fixes into my DLC environment:
a. cuDNN v. 8.2.1 (critical solution)
b. tensorflow 2.1.0 (this was already installed for me)

With the above combination, the GPU is recognized in the testscrip.py output (tensorflow doesn't look for missing v11 libraries - I am not  clear how installing cuDNN v 8.2.1 solves this, maybe the path is updated).   For those new folks who don't know what test script output is, if successfully run and done, you will find in a TEST-Alex-yr-mo-day/videos folder a 1 second long video called "reachingvideo1shortDLC_mobnet_35_TEST<date>shuffle1_5_labeled.mp4". Here you will see an untrained AI labeling points on video frames - in my case all points migrating towards some moving whiskers of the mouse.

Another tidbit for those following directions in the instructions as if you're expected to run the commands on the command line: It is much easier to do all of it in Jupyter Notebook except for resetting the kernel during training.

1. Run Anaconda Navigator as admin (type in windows search select right caret > run as admin)
2. Change in middle dropdown menu: base (root) to DEEPLABCUT
3. Install Jupyter Lab (notebook should already be installed), and run it (appears in browser). I prefer to run from the Anaconda3 commandline: "jupyter notebook" or "jupyter lab". 
4. Copy from DeepLabCut/examples/JUPYTER folder the "Demo_labeledexample_MouseReaching.ipynb" file to the DeepLabCut/examples folder.
5. Open the new notebook copied file and use the DEEPLABCUT python kernel.  
6. Run each step individually (shift-Enter), including up to training.
a. Looking at labeled-data folder for Reaching-Mackenzie-2018-08-30 will have images showing what Mackenzie labelled for bodyparts of the hand, the lever and tongue.
7. If after running the training step, all goes well the GPU should spin up and trained kernel will save every 300 iterations (I stopped after 10 saves or 3000 iterations)
a. I had trouble stopping the kernel in the notebook: "Stop" button and restart did not restart the kernel. Instead I had to choose "Close and halt", and from the jupyter notebook file chooser, re-open the Demo_*.ipynb notebook, restarting the kernel.
b. After doing a), I re-ran all the steps except skipping the training step.
8. You should now be able to analyze videos, instead I switched the test video to the non-training video by changing  videofile_path = os.path.join(os.getcwd(),'Reaching-Mackenzie-2018-08-30/videos/MovieS2_Perturbation_noLaser_compressed.avi')  
9. Running all the steps you will have a new video in the Reaching-Mackenzie-2018-08-30 videos folder called "MovieS2_Pertubation_noLaser..." etc. showing the successful labelling of a novel video.

Good luck!
